### DISCUSSION TITLE
查询的时候将collection加载到内存，如果特别多的话不会导致内存溢出吗

### DISCUSSION THREAD

**Discussion by rilweic** [milvus-io/milvus#18004] [Q&A and General discussion]
<img width="937" alt="image" src="https://user-images.githubusercontent.com/9164714/176863637-45181baa-7622-4bd7-8cc4-7faec88042db.png">

**Comment by yhmo** [comment]
collection数据太多了内存不够的时候load()会报错，目前要求数据都在内存里才能search。以后应该会有优化方案

**Reply by zsinba** [reply]
> collection数据太多了内存不够的时候load()会报错，目前要求数据都在内存里才能search。以后应该会有优化方案

至今还没有优化方案。 
随着文档的增加， 向量的数据增加， 内存在直线上升。
等等解决方案。

**Reply by yhmo** [reply]
mmap是一个方案，就看你容不容忍性能的下降。

**Reply by zsinba** [reply]
> mmap

有测试过下降了多少吗？ 当前全内存方案，查询时间 在 200ms 左右完成整个请求。

**Reply by yhmo** [reply]
下降多少跟内存密切相关，内存量是数据量的1/4左右时，性能下降不多还能忍受，内存再少的话就损失很大了，自己体验一下就好。
实际上数据全都在内存里是最快的，有磁盘交换就会慢很多，理想的多快好省不存在的。

**Reply by zsinba** [reply]
> 下降多少跟内存密切相关，内存量是数据量的1/4左右时，性能下降不多还能忍受，内存再少的话就损失很大了，自己体验一下就好。 实际上数据全都在内存里是最快的，有磁盘交换就会慢很多，理想的多快好省不存在的。

对Milvus 的这一点了解有限，Mysql 百 G 的存储，也可以快速返回，也就是几个 G 的内存占用，根据需要从硬盘捞数据，不见得有什么影响。
是因为Milvus 要做向量计算，所以要在内存里等着吗？

**Reply by xiaofan-luan** [reply]
向量索引是图索引，几乎无法cache。而绝大多数标量存储都是顺序存储。

目前milvus也提供了基于磁盘的方案，但这依然依赖于超高性能的存储（NVMe SSD）

**Reply by yhmo** [reply]
对于高维向量，每个维度是float32，4个字节。1024维就是4KB。
现在有1000万条1024维的向量，数据量是40GB。它们存储在磁盘里，现在我要输入一条1024维向量，跟这1000万条向量分别计算欧式距离，然后返回距离最小的10条结果。我给你4G内存，你会怎样调度使得单次查询的latency控制在10毫秒以下？先加载4GB(100万条向量)，计算完后再加载另外100万条，如此循环10次？什么硬盘能做到10毫秒？

IVF/HNSW这些内存索引，可以做到10毫秒。之所以这些索引要全加载进内存，是因为查询时命中哪片区域无法预测，如果做成磁盘交换的，根本无法稳定高速地查询。
DISKANN是依赖硬盘的索引，但它也要加载压缩后的向量进内存，内存中先做粗略检索，定位到硬盘的某些位置把所需的向量读取上来再做精确搜索，它的性能也仅能达到百毫秒级，而且对硬盘性能要求很高。

**Reply by zsinba** [reply]
那这不是无解了吗？ 
1000W --40G
2000W -- 80G
4000W --160G
只能做服务器的横向扩展来解决数据增量吗？

**Reply by zsinba** [reply]
感觉算的不对呀， 1024 维是 4KB； 如果分片大小是 1000 个字符； 
那 20 万字占内存
`# Constants
num_segments = 200
memory_per_segment_bytes = 4096

# Total memory in bytes
total_memory_bytes = num_segments * memory_per_segment_bytes

# Convert bytes to MB
total_memory_mb = total_memory_bytes / 1048576
total_memory_mb
`
存储一个 20 万字的文章，每 1000 个字符作为一个分片，并且每个分片的向量维度为 1024 所需的内存约为 0.78125 MB.
但实际上感觉比这大很多呀

**Reply by yhmo** [reply]
我只针对向量说的。文章多少个字符切成一块，由使用者自己决定，每条向量的维度由模型决定。你也可以选择1万个字符切一片，向量维度256，这跟我说的例子并无关系。

**Reply by yhmo** [reply]
数据只增不减，消耗的资源肯定也是只增不减。如果你每次都做全量的检索，而增量无上限，那你应该考虑一下你的业务模式是不是该调整一下了。地球现在能承载80亿人口，那它以后能承载一万亿么？

**Reply by zsinba** [reply]
> 数据只增不减，消耗的资源肯定也是只增不减。如果你每次都做全量的检索，而增量无上限，那你应该考虑一下你的业务模式是不是该调整一下了。地球现在能承载80亿人口，那它以后能承载一万亿么？

谢谢
